packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -376,5 +400,39 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function targetNodeFunctionAncestorNodeHasVoidReturnType(
+      node: TSESTree.Node,
+    ): boolean {
+      const targets: [
+        AST_NODE_TYPES.FunctionDeclaration,
+        AST_NODE_TYPES.ArrowFunctionExpression,
+        AST_NODE_TYPES.FunctionExpression,
+      ] = [
+        AST_NODE_TYPES.FunctionDeclaration,
+        AST_NODE_TYPES.ArrowFunctionExpression,
+        AST_NODE_TYPES.FunctionExpression,
+      ];
+
+      if (!node.parent) {
+        return false;
+      }
+
+      if (
+        targets[0] === node.parent.type ||
+        targets[1] === node.parent.type ||
+        targets[2] === node.parent.type
+      ) {
+        if (
+          node.parent.returnType?.typeAnnotation.type ===
+          AST_NODE_TYPES.TSVoidKeyword
+        ) {
+          return true;
+        }
+        return false;
```

Comment:
[Bug] I just noticed that we determine whether function signature has `void` return type or not by doing pure syntax check. 

Thus, the following case is false positive (the rule shouldn't report error, since function returns `Foo` - an alias of `void` type):

```ts
type Foo = void

function test(): Foo {
  return console.log()
}
```

[playground](https://deploy-preview-8632--typescript-eslint.netlify.app/play/#ts=5.3.3&showAST=types&fileType=.ts&code=C4TwDgpgBAYg9nKBeKA3OBLAJgKBwMwFcA7AY2AzmKmAgGdgAKASgC5YEoBvHKKAJwjBC-aqSp04AGwgA6KXADmLHAF8cQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90YANrhsORNGgdokADSatWLdkiVhnRQDVJUgJJN701fjiIT2AL7eAupp%2BPkA&tsconfig=&tokens=false)